crates/astria-core/src/protocol/transaction/v1alpha1/action_group/mod.rs

```diff
@@ -0,0 +1,201 @@
+    (Ics20Withdrawal, ActionGroup::General),
+    (InitBridgeAccountAction, ActionGroup::UnbundleableGeneral),
+    (BridgeLockAction, ActionGroup::General),
+    (BridgeUnlockAction, ActionGroup::General),
+    (BridgeSudoChangeAction, ActionGroup::UnbundleableGeneral),
+    (FeeChangeAction, ActionGroup::Sudo),
+    (FeeAssetChangeAction, ActionGroup::Sudo),
+    (IbcRelay, ActionGroup::General),
+    (IbcSudoChangeAction, ActionGroup::UnbundleableSudo),
+);
+
+impl Action {
+    const fn group(&self) -> ActionGroup {
+        match self {
+            Action::Sequence(_) => SequenceAction::GROUP,
+            Action::Transfer(_) => TransferAction::GROUP,
+            Action::ValidatorUpdate(_) => ValidatorUpdate::GROUP,
+            Action::SudoAddressChange(_) => SudoAddressChangeAction::GROUP,
+            Action::IbcRelayerChange(_) => IbcRelayerChangeAction::GROUP,
+            Action::Ics20Withdrawal(_) => Ics20Withdrawal::GROUP,
+            Action::InitBridgeAccount(_) => InitBridgeAccountAction::GROUP,
+            Action::BridgeLock(_) => BridgeLockAction::GROUP,
+            Action::BridgeUnlock(_) => BridgeUnlockAction::GROUP,
+            Action::BridgeSudoChange(_) => BridgeSudoChangeAction::GROUP,
+            Action::FeeChange(_) => FeeChangeAction::GROUP,
+            Action::FeeAssetChange(_) => FeeAssetChangeAction::GROUP,
+            Action::Ibc(_) => IbcRelay::GROUP,
+            Action::IbcSudoChange(_) => IbcSudoChangeAction::GROUP,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ActionGroup {
+    General,
+    UnbundleableGeneral,
+    Sudo,
+    UnbundleableSudo,
+}
+
+impl ActionGroup {
+    pub(super) fn is_bundleable(self) -> bool {
+        matches!(self, ActionGroup::General | ActionGroup::Sudo)
+    }
+
+    pub(super) fn is_sudo(self) -> bool {
+        matches!(self, ActionGroup::Sudo)
+    }
+}
+
+impl fmt::Display for ActionGroup {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ActionGroup::General => write!(f, "general"),
+            ActionGroup::UnbundleableGeneral => write!(f, "unbundleable general"),
+            ActionGroup::Sudo => write!(f, "sudo"),
+            ActionGroup::UnbundleableSudo => write!(f, "unbundleable sudo"),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error(transparent)]
+pub struct Error(ErrorKind);
+
+impl Error {
+    #[must_use]
+    pub fn kind(&self) -> &ErrorKind {
+        &self.0
+    }
+
+    fn mixed(
+        original_group: ActionGroup,
+        additional_group: ActionGroup,
+        action: &'static str,
+    ) -> Self {
+        Self(ErrorKind::Mixed {
+            original_group,
+            additional_group,
+            action,
+        })
+    }
+
+    fn not_bundleable(group: ActionGroup) -> Self {
+        Self(ErrorKind::NotBundleable {
+            group,
+        })
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ErrorKind {
```

Comment:
private? Do you need to check why bundling failed and take action based on it?

Reply:
Responded on the comment above